k256/src/ecdsa.rs

```diff
@@ -213,7 +213,7 @@ impl SignPrimitive<Secp256k1> for Scalar {
         // Compute 𝒔 as a signature over 𝒓 and 𝒛.
         let s = k_inv * (z + (r * self));
 
-        if s.is_zero().into() {
+        if s.is_zero().into() || r.is_zero().into() {
```

Comment:
FWIW, these checks duplicate the checks here:

https://github.com/RustCrypto/signatures/blob/800eda2/ecdsa/src/lib.rs#L206-L216

Reply:
I opened https://github.com/RustCrypto/signatures/pull/730 which hopefully clarifies both the logic and documentation around checks that `r` and `s` are non-zero.

Note that these checks are part of the invariant of the `ecdsa::Signature` type and the constructors return an error if either component is zero.

Reply:
There's actually a lot of duplicated logic there which would be nice to merge somehow.

But I would disagree that these checks are redundant just because `Signature::from_scalars` implements them. It is a part of the public API, and it makes sense for it to preserve that invariant too.

Reply:
> There's actually a lot of duplicated logic there which would be nice to merge somehow.

I think I'd like to extract some functions like `ecdsa::hazmat::{sign_raw, verify_raw}`... the duplication largely exists to implement low-S normalization, but if it can call a low-level function we don't need to duplicate the rest of the implementation. See also #908 for a verification-side analogue of this problem (where the generic implementation *is* being used, but low-S normalization is no longer honored)

> But I would disagree that these checks are redundant just because Signature::from_scalars implements them. It is a part of the public API, and it makes sense for it to preserve that invariant too.

If it makes the code easier to audit I'm fine with it.